src/AutoRest.CSharp/Common/Output/Models/Types/HelperTypeProviders/ChangeTrackingListProvider.cs

```diff
@@ -0,0 +1,282 @@
+            var iReadOnlyListSignature = new ConstructorSignature(Type, null, null, MethodSignatureModifiers.Public, new Parameter[] { iReadOnlyListParam });
+            var iReadOnlyListVariable = new ParameterReference(iReadOnlyListParam);
+            var iReadOnlyListbody = new MethodBodyStatement[]
+            {
+                new IfStatement(Snippets.NotEqual(iReadOnlyListVariable, Snippets.Null))
+                {
+                    new AssignValueStatement(_innerList, Snippets.Linq.ToList(iReadOnlyListVariable))
+                }
+            };
+
+            yield return new Method(iReadOnlyListSignature, iReadOnlyListbody);
+        }
+
+        protected override IEnumerable<CSharpType> BuildTypeArguments()
+        {
+            yield return _t.Type;
+        }
+
+        protected override IEnumerable<CSharpType> BuildImplements()
+        {
+            yield return new CSharpType(typeof(IList<>));
+            yield return new CSharpType(typeof(IReadOnlyList<>));
+        }
+
+        protected override IEnumerable<FieldDeclaration> BuildFields()
+        {
+            yield return _innerlListField;
+        }
+
+        protected override IEnumerable<PropertyDeclaration> BuildProperties()
+        {
+            yield return new PropertyDeclaration(null, MethodSignatureModifiers.Public, typeof(bool), "IsUndefined", new ExpressionPropertyBody(Snippets.Equal(_innerList, Snippets.Null)));
+            yield return BuildCount();
+            yield return BuildIsReadOnly();
+            yield return BuildIndexer();
+        }
+
+        private PropertyDeclaration BuildIsReadOnly()
+        {
+            return new PropertyDeclaration(null, MethodSignatureModifiers.Public, typeof(bool), "IsReadOnly",
+                        new ExpressionPropertyBody(new TernaryConditionalOperator(
+                            new MemberExpression(Snippets.This, "IsUndefined"),
+                            Snippets.False,
+                            new MemberExpression(new InvokeInstanceMethodExpression(Snippets.This, _ensureListSignature), "IsReadOnly"))));
+        }
+
+        private PropertyDeclaration BuildCount()
+        {
+            return new PropertyDeclaration(null, MethodSignatureModifiers.Public, typeof(int), "Count",
+                new ExpressionPropertyBody(new TernaryConditionalOperator(
+                    new MemberExpression(Snippets.This, "IsUndefined"),
+                    Snippets.Literal(0),
+                    new MemberExpression(new InvokeInstanceMethodExpression(Snippets.This, _ensureListSignature), "Count"))));
+        }
+
+        private PropertyDeclaration BuildIndexer()
+        {
+            return new PropertyDeclaration(null, MethodSignatureModifiers.Public, _t.Type, "this", new MethodPropertyBody(
+                new MethodBodyStatement[]
+                {
+                    new IfStatement(new BoolExpression(new MemberExpression(Snippets.This, "IsUndefined")))
```

Comment:
there is a possible refactor here to improve the readability here:
```suggestion
                    new IfStatement(new BoolExpression(IsUndefinedProperty))
```
if we define this `IsUndefined` property as a property of `PropertyDeclaration` in this class, and we should add an implicit cast operator from `PropertyDeclaration` to `ValueExpression` as we already have for `FieldDeclaration`.